feat: add setup guide for dense vector search

rfazeli opened the pull request:
This PR adds a setup guide for adding dense vector search (knn search) in Elasticsearch.

Signed off by: reza@ibm.com

sne3091 commented on integrations/extensions/docs/elasticsearch-install-and-setup/text_embedding_deploy_and_use.md:
````diff
@@ -0,0 +1,184 @@
+  export ES_USER=<username>
+  export ES_PASSWORD=<password>
+  export ES_CACERT=<path-to-your-cert>
+  ```  
````
To keep it consistent, in general, it would be better to export the names of the indices, pipelines here as well? They can then be used consistently later on , all through the process

jwm4 replied:
+1

rfazeli replied:
Good point! I wonder if it'd be better to do this as part of another PR for all the docs we have to keep them consistent.

jwm4 reviewed (changes requested):
Still working on the wording of the disclaimer regarding E5.

zach-shu commented on integrations/extensions/starter-kits/elasticsearch/README.md:
````diff
@@ -66,6 +66,23 @@ but you will need to find set-up instructions appropriate to that environment.
     ELSER v2 is only available for the 8.11 version of Elasticsearch
     * Compound search  
       You can combine different types of queries in a compound query. Learn more about it from this [Elasticsearch tutorial](https://www.elastic.co/guide/en/elasticsearch/reference/8.10/semantic-search-elser.html#text-expansion-compound-query).
+  * To use dense vector search (k-nearest neighbours search), you need to set `knn_body` as a session variable and set the `knn` variable to `knn_body` as shown below:
+    * Here is an example knn body you can use when setting up the `knn_body` session variable
+    ```json
+      {
+        "field": "text_embedding.predicted_value",
+        "query_vector_builder": {
+          "text_embedding": {
+            "model_id": "elastic__multilingual-e5-small-optimized",
+            "model_text": "how to set up custom extension?"
+          }
+        },
+        "k": 10,
+        "num_candidates": 100
+      }
+      ```
+      <img src="assets/use_elasticsearch_custom_extension_knn.png" width="669" height="627" />
+  
````
Setting `knn_body` as a session variable is not required, so it'd be better to replace `you need to` with `you can`.

zach-shu replied:
This section about dense vector search will be better to be placed before `Compound Search`

zach-shu reviewed (changes requested):
Thank you for creating this guide, Reza! I was able to make it work in CP4D as well following the instructions here. And I have a couple of comments below.

jwm4 commented on integrations/extensions/docs/elasticsearch-install-and-setup/text_embedding_deploy_and_use.md:
````diff
@@ -0,0 +1,184 @@
+```bash
+$ python -m pip install "eland[pytorch]"
+```
+This library allows us to pull and deploy a 3rd-party text embedding model to our Elasticsearch instance.
````
Charles Quincy and I worked together on the following IP warning for 3rd party embedding models.  We are both comfortable with this wording:

> CAUTION: Open source and 3rd party models are not in scope of IBM or Elastic indemnity clauses. Customers must accept relevant terms and conditions to choose or bring their own models. Additionally, IBM has not assessed Elastic's supported multi-lingual models so any use of Elastic-supported models should be understood thoroughly both respect to the terms of use for those models and the terms of use of all of the data that was used to train those models.

rfazeli replied:
Perfect thanks. I'm assuming `understood thoroughly both respect to the terms` is `understood thoroughly both with respect to the terms`

jwm4 replied:
Right, good catch -- yes, please include that fix.

jwm4 reviewed (approved):
My main concern is addressed.  With that said, I haven't done a full review, so I would wait for at least one other approval before anyone merges this.

rfazeli commented:
Thanks everyone for your reviews. @zach-shu I've addressed your comments. @sne3091 I think it might be best to address setting index and pipeline names as environment variables across all guides in a separate PR.

zach-shu commented on integrations/extensions/docs/elasticsearch-install-and-setup/text_embedding_deploy_and_use.md:
````diff
@@ -0,0 +1,189 @@
+## Install the eland library
+Run the command below to install the [eland](https://github.com/elastic/eland) library.
+```bash
+python -m pip install "eland[pytorch]"
````
Hi Reza, I think it would be nice to add something more explicit about compatibility. For example, 

NOTE: As of the time this documentation was written, `eland` only supports Python 3.8, 3.9, and 3.10. Please refer to the eland library [compatibility section](https://github.com/elastic/eland?tab=readme-ov-file#compatibility) to make sure you're using compatible Python and Elasticsearch versions. 

Also, adding the note right after this Python command looks more intuitive to me.

rfazeli replied:
Good point, thanks! Will add this note

eric-wayne commented on integrations/extensions/docs/elasticsearch-install-and-setup/text_embedding_deploy_and_use.md:
````diff
@@ -0,0 +1,189 @@
+  ```  
+You can find the credentials from the service credentials of your Elasticsearch instance.
+## Pull and deploy an embedding model
+Run the command below to pull your desired model from the [Huggingface Models Hub](https://huggingface.co/models) and deploy it on your Elasticsearch instance:
````
For some reason, I was not able to successfully run eland_import_hub_model after running python -m pip install "eland[pytorch]".  I'm getting this error: `command not found: eland_import_hub_model`

I was able to run the equivalent docker command.  But I think it ran into the issue that since I'm using the same ES ICD instance, the model was already deployed.

eric-wayne commented on integrations/extensions/docs/elasticsearch-install-and-setup/text_embedding_deploy_and_use.md:
````diff
@@ -0,0 +1,189 @@
+{"task":<task-id>}
+```
+
+The reindexing process can take around 10 minutes. You can use the task id that is returned above to check the status of the process.
````
Is there a specific part of the response of the `_tasks` call that the reader should look for?  I guess it's just `"completed":true` ?

rfazeli replied:
Yes, that's correct. I'll add a note pointing that out.

eric-wayne commented on integrations/extensions/docs/elasticsearch-install-and-setup/text_embedding_deploy_and_use.md:
```diff
@@ -0,0 +1,189 @@
+Once the process is complete, you should see `wa-docs-with-embeddings` in the list of your indices http://localhost:5601/app/enterprise_search/content/search_indices
+
+## Run semantic search
+After the dataset has been enriched with vector embeddings, you can query the data using semantic search. 
```
I'm getting an error running this. I'll pause my review here.

`{"error":{"root_cause":[{"type":"query_shard_exception","reason":"failed to create query: [knn] queries are only supported on [dense_vector] fields","index_uuid":"E50WgDeoS6ONscKjyHkO3Q","index":"ewayne-wa-docs-with-embeddings"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"dfs","grouped":true,"failed_shards":[{"shard":0,"index":"ewayne-wa-docs-with-embeddings","node":"usod4PNbQU6r7LrzoShkpw","reason":{"type":"query_shard_exception","reason":"failed to create query: [knn] queries are only supported on [dense_vector] fields","index_uuid":"E50WgDeoS6ONscKjyHkO3Q","index":"ewayne-wa-docs-with-embeddings","caused_by":{"type":"illegal_argument_exception","reason":"[knn] queries are only supported on [dense_vector] fields"}}}]},"status":400}`

rfazeli replied:
This is interest! I am able to replicate the error. This happens when [wa-docs-100](http://localhost:5601/app/enterprise_search/content/search_indices/wa-docs-100/) is used as the source index. I didn't run into this because I used the original [wa-docs](http://localhost:5601/app/enterprise_search/content/search_indices/wa-docs/) index as the source. 

The only difference I see between these two indices is that in wa-docs all columns are type `text` while in wa-docs-100 the title column is automatically set to the `keyword` type and I can't seem to modify it. @sne3091 @zach-shu do you know how wa-docs-100 was created?

![image](https://github.com/watson-developer-cloud/assistant-toolkit/assets/5916032/239c0e47-3801-4d4d-9e04-46e0bfa63716)

![image](https://github.com/watson-developer-cloud/assistant-toolkit/assets/5916032/f5753901-e6dc-4d93-b4aa-0dfdc2900b2c)

rfazeli replied:
This issue is not encountered anymore so closing this.

sne3091 commented:
> . @sne3091 I think it might be best to address setting index and pipeline names as environment variables across all guides in a separate PR.

On a call, couple weeks ago including @jwm4 , we discussed that we would utilize this practice for consistency moving forward as to avoid naming indices/pipelines etc. and make it more generic and both @zach-shu 's PR and mine incorporated that feedback. 

It would be great to maintain that consistency now and moving forward for all the guides we write up as opposed to coming back to visit later on since its just becomes yet another thing to revisit later on in the growing number of guides. So, I'd recommend including that change now but up to you

rfazeli commented:
Thanks, @sne3091! I think we still have a few PRs where env variables are not included but I went ahead and used env variables for index names and ingest pipelines here.

> On a call, couple weeks ago including @jwm4 , we discussed that we would utilize this practice for consistency moving forward as to avoid naming indices/pipelines etc. and make it more generic and both @zach-shu 's PR and mine incorporated that feedback.

> It would be great to maintain that consistency now and moving forward for all the guides we write up as opposed to coming back to visit later on since its just becomes yet another thing to revisit later on in the growing number of guides. So, I'd recommend including that change now but up to you

zach-shu reviewed (approved):
Looks good to me now. Thanks!
